linalg/ode.hpp

```diff
@@ -223,37 +232,61 @@ class RK8Solver : public ExplicitRKSolver
 };
 
 
-/** An explicit Adams-Bashforth method. */
-class AdamsBashforthSolver : public ODESolver
+/** An Linear Multi Step base class. */
+class LMSSolver : public ODESolver
 {
-private:
-   int s, smax;
-   const double *a;
-   Vector *k;
+protected:
```

Comment:
Do all these member variables need to be made protected?

Reply:
I've had a little more time to read through this PR more thoroughly and am at the understanding that the scope of the originally private variables in `AdamsBashforthSolver` has been broadened to protected in `LMSSolver` so they can be manipulated by the classes that derive `LMSSolver`.  Let me know if that isn't correct.

If that is correct, it's not clear to me why `LMSSolver` as-is should inherit from `ODESolver` (e.g., is there an instance where an `LMSSolver` object needs to be passed as or cast to an `ODESolver` object?). Would it make sense to have the `LMSSolver` class be a struct, perhaps named something like `LMSData`, that is then a member variable of `AdamsBashforthSolver` and `AdamsMoultonSolver` (which would revert to inheriting from `ODESolver`)? Such an approach would avoid adding another abstract class to the inheritance tree.

I'll admit that as a proponent of "has-a" over "is-a", I may be in the minority of the MFEM community (i.e., there a plenty of instances in the MFEM code where an abstract class layer is added for ease of reusing the implementations in the abstract class by the derived classes), but I figured I'd voice my opinion as an assigned reviewer.

Reply:
I see this has been marked resolved... was there a corresponding code change that I missed or is the marking resolved to be taken as a "the author of the PR disagrees that a change should be discussed"?

Reply:
For what it's worth, I will take no offense if the resolved marking was to be taken as "the author of the PR rejects the proposed change/discussion"... I'm more so asking because if that is indeed how the meaning of the resolved marking, then I will unsubscribe from the PR (which will then be presumably reviewed by the other two reviewers).

Reply:
Test...

Reply:
> The LMSSolver can be used in all location where a ODESolver can be used

I don't see where the `LMSSolver` implements the pure virtual `ODESolver::Step`, and thus cannot be instantiated as an `ODESolver`... i.e., it can only be instantiated through a derived class.

Reply:
> In ODESolver* ODESolver::Select(int ode_solver_type)
the ABSolvers and AMSolvers can be selected and are returned as *ODESolver.

> Doing this by making LMSSolver an ODESolver seemed the obvious thing for me.

> Therefor the code reuse argument won out in my personal consideration.

I am of the opinion that inheritance should not be used solely for code reuse and should instead be restricted to cases where polymorphism is needed, and having the ABSolvers and AMSolvers inherit from `ODESolver` meets the polymorphism need without introducing another layer of inheritance that doesn't have a polymorphism need (e.g., the ABSolvers and AMSolvers never need to be passed as an `LMSSolver`)

Reply:
> However, the ODESolver class already has GetStateVector and SetStateVector members in the particular ODESolvers these needs to be linked with the LMSData class. Which for me passes then the point of creating more obfuscation then solving it (by eliminating the abstract class).

I may be missing something but I believe the linking mentioned would be something along the lines of 

```c++
void AdamsMoultonSolver::SetStateVector(int i, Vector &state) override
{
  lmsData.SetStateVector(i, state);
}
```

To me, this is preferable over class inheritance that imposes a particular implementation on all derived classes.  As an example, the original `AdamsMoultonSolver` and `AdamsBashSolver` classes encapsulated the implementation of stage storage, so any changes to the choice of stage storage would be limited to the class.

Reply:
> Perhaps one or two (dt_?) could be made private. But I personally do not mind the flexibility for future changes

I worry about expanding the scope of variables for the sake of flexibility.  If a derived class changes `_dt` for example, will `LMSSolver::CheckTimestep` still work?  If the `LMSSolver` member functions make assumptions that protected member variables are only modified by `LMSSolver` member functions, I believe that should be make private and only accessed by derived classes through read-only accessors.

Reply:
Thank you for reengaging with the PR. I'll have a second look at the LMSData suggestion, because you make  valid points that it is not essential, but merely `useful` (aka enables laziness). 

Perhaps certain parts of the current LMSSolver can be moved to a data struct and other parts can be moved to the ODESOlver base class.
In hind sight, I might have put not changing ODESolver to high on the wish-list.

Reply:
I re-implemented the suggested` LMSData` struct. To generalize it a bit a named it `StateData`.  I included the data member to the base class. Therefore there is only one place were the member functions are passed trough. This also allowed the Genalpha time steppers to use the StateData. The latter was not necessary but makes things more uniform.

I removed the `CheckTimestep` function and put the functionality in the AB and AM `step` routine. This resulted in duplication of this trivial code. This avoids the suggested confusion. 

`dt_` could be renamed `dt_prev` if desired.. Or a comment could be added.